### PR TITLE
AMP web push Chrome 62 notification API changes and other improvements

### PR DESCRIPTION
**Pull request by jasonpang**
### Compatibility with upcoming Chrome 62 notification API changes

Around October 17 (2 weeks and 5 days from this PR open date), Chrome 62 will launch with the following new notification API behavior:

  - [Accessing notification permissions within a secure cross-origin IFrame will return denied instead of default for default permissions](https://bugs.chromium.org/p/chromium/issues/detail?id=765531)
    - Example: An IFrame `https://site.com` is embedded under `https://www.google.com/s/amp/site.com`. Querying the notification permission of `https://site.com` while inside the IFrame will return `denied` if you've never granted or denied permissions. Prior to Chrome 62, the query would have returned `default`.
    - Reason for Change:
      - [Notification permissions can no longer be requested from iframes](https://www.chromestatus.com/features/6451284559265792) (deprecated in Chrome 59, removed in Chrome 62)
      - Whether notification permissions can be requested from an iframe seems unrelated to accessing the iframe's actual notification permission state, but in fact [the notification permission state returns the result of requesting permissions without user action](https://bugs.chromium.org/p/chromium/issues/detail?id=765531#c2).

Result of permission state query from within a secure cross-origin iframe:

| Actual Permission | Chrome 61- Returns | Chrome 62+ Returns |
| --- | --- | --- |
| Granted | Granted | Granted | 
| Denied | Denied | Denied |
| Default | Default | **Denied**  |

The current release of AMP web push queries the canonical origin's (i.e. iframe's) notification permission in deciding to present subscription or unsubscription options.

**After Chrome 62 launches, our current AMP web push extension will treat all new visitors as having blocked notification permissions**, because new visitors never having granted or denied permissions would have a `default` permission, and Chrome 62 will return `denied` when `default`.

Our workaround is:
  - Assume new users without any storage information (see below) have `default` permissions, despite permission queries returning `denied` permissions
  - When the user subscribes to push, store the actual permission in the canonical origin's LocalStorage
    - Due to browser requirements, visitors always subscribe to  AMP web push in a top-level frame (a popup or redirect)
    - The ambiguous permission behavior described above only applies to cross-origin iframes; secure top-level frames can access the true notification permission state without restrictions
  - The next time, if we detect a stored notification permission of blocked, show our blocked permission page section
    - Publishers can define separate page sections to be shown based on the user's subscription state; the blocked page section is one of these

It's possible for a user blocking notification permissions to clear storage and attempt to subscribe; this user will see a message telling them to unblock permissions and an X button to quit. If they unblock permissions, the regular subscription flow will resume.

**So, in summary, for this commit:**

  - **The initial permission check now accounts for Chrome 62's permission change by using additional state**
  - **The subscription popup now accounts for users blocking permissions**

### Other improvements

- **When subscribing or unsubscribing, disable the button until the process is complete or canceled**
   
  Previously, users could click the button multiple times.

- **Relax HTTPS restriction to allow testing on HTTP localhost**

  AMP web push requires HTTPS to be able to subscribe the user to push. Previously, we checked whether the URL begain with `https://`. However, Chrome treats `http://localhost` and `http://127.0.0.1` as secure, and users can subscribe to push on these origins. Therefore, the HTTPS check has been increased to allow these two origins as well.

- **Make installed service worker URL check smarter**

  If a service worker is already registered, AMP web push checks whether it is the same one specified in the extension config. Previously, this comparison was an exact string match. Now, everything but the query parameters have to match exactly, and query parameters can be out of order, or the installed worker can have extra query parameters.  

## Pull Request Tasks

- [ ] **Review use of `localStorage`**

  `gulp presubmit` flagged my use of `localStorage`. The need for storage is explained in the first section of Chrome 62's permission changes.

### CODE REVIEW

File: extensions/amp-web-push/0.1/web-push-service.js [hunk 1]
```diff
@@ -158,19 +153,25 @@ export class WebPushService {
...
+    iframeLoadPromise
+        .then(() => {
+          dev().fine(
+              TAG,
+              `Helper frame ${this.config_['helper-iframe-url']} ` +
+            'DOM loaded. Connecting to the frame via postMessage()...'
+        );
+          return this.frameMessenger_.connect(
+              this.iframe_.getDomElement().contentWindow,
+              new URL(this.config_['helper-iframe-url']).origin
```
Comment (dvoytenko) [inline extensions/amp-web-push/0.1/web-push-service.js]: `URL` availability could be a problem here, right? Use `parseUrl`?

File: extensions/amp-web-push/0.1/web-push-service.js [hunk 2]
```diff
@@ -494,26 +613,86 @@ export class WebPushService {
...
+            default:
+            /*
+            Because notification permissions are not granted, we need to open a
+            popup asking the user to grant permissions.
+           */
+              const permissionDialogWindow = this.openPopupOrRedirect();
+              this.checkPermissionDialogClosedInterval_(
+                  permissionDialogWindow,
+                  onPopupClosed
+            );
```
Comment (dvoytenko) [inline extensions/amp-web-push/0.1/web-push-service.js]: nit: +2

**Comment by dvoytenko** [comment]
@jasonpang Ready for merge? I saw commit and things look all fixed, right?

**Comment by jasonpang** [comment]
Hey @dvoytenko,

All tested and ready to merge!

**Comment by dvoytenko** [comment]
@jasonpang Thanks! Merging.

**Pull request merged**